ext/nnstreamer/tensor_filter/tensor_filter_tizen_hal.cc

```diff
@@ -0,0 +1,326 @@
+ *
+ * This is the per-NN-framework plugin (Tizen HAL) for tensor_filter.
+ */
+
+#include <iostream>
+#include <string>
+
+#include <glib.h>
+#include <nnstreamer_cppplugin_api_filter.hh>
+#include <nnstreamer_log.h>
+#include <nnstreamer_plugin_api_util.h>
+#include <nnstreamer_util.h>
+
+#include <hal-ml.h>
+
+namespace nnstreamer
+{
+namespace tensor_filter_tizen_hal
+{
+extern "C" {
+void init_filter_tizen_hal (void) __attribute__ ((constructor));
+void fini_filter_tizen_hal (void) __attribute__ ((destructor));
+}
+
+/** @brief tensor-filter-subplugin concrete class for Tizen HAL */
+class tizen_hal_subplugin final : public tensor_filter_subplugin
+{
+  private:
+  static const char *fw_name;
+  static tizen_hal_subplugin *registeredRepresentation;
+  static const GstTensorFilterFrameworkInfo framework_info;
+  hal_ml_h hal_handle;
+  gchar *backend_name;
+
+  public:
+  static void init_filter_tizen_hal ();
+  static void fini_filter_tizen_hal ();
+
+  tizen_hal_subplugin ();
+  ~tizen_hal_subplugin ();
+
+  tensor_filter_subplugin &getEmptyInstance ();
+  void configure_instance (const GstTensorFilterProperties *prop);
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+};
+
+const char *tizen_hal_subplugin::fw_name = "tizen-hal";
+
+/**
+ * @brief Constructor for tizen_hal_subplugin.
+ */
+tizen_hal_subplugin::tizen_hal_subplugin ()
+    : tensor_filter_subplugin (), hal_handle (nullptr), backend_name (nullptr)
+{
+}
+
+/**
+ * @brief Destructor for Tizen HAL subplugin.
+ */
+tizen_hal_subplugin::~tizen_hal_subplugin ()
+{
+  if (backend_name) {
+    g_free (backend_name);
+    backend_name = nullptr;
+  }
+
+  if (hal_handle) {
+    hal_ml_destroy (hal_handle);
+    hal_handle = nullptr;
+  }
+}
+
+/**
+ * @brief Method to get empty object.
+ */
+tensor_filter_subplugin &
+tizen_hal_subplugin::getEmptyInstance ()
+{
+  return *(new tizen_hal_subplugin ());
+}
+
+/**
+ * @brief Method to prepare/configure Tizen HAL instance.
+ */
+void
+tizen_hal_subplugin::configure_instance (const GstTensorFilterProperties *prop)
+{
+  /* parse backend from custom prop */
+  if (!prop->custom_properties)
+    throw std::invalid_argument ("Custom properties are not given. Provide a valid backend.");
+
+  /* prop->custom_properties has value of "backend:backend_name,key1:val1,key2:val2" */
+  gchar **options = g_strsplit (prop->custom_properties, ",", -1);
+
+  for (guint op = 0; op < g_strv_length (options); ++op) {
+    gchar **option = g_strsplit (options[op], ":", -1);
+
+    if (g_strv_length (option) >= 2) {
+      g_strstrip (option[0]);
+      g_strstrip (option[1]);
+
+      if (g_ascii_strcasecmp (option[0], "backend") == 0) {
+        backend_name = g_strdup (option[1]);
+        break;
```

Comment:
I think break is unnecessary here (printing warning message for useless option?)

Reply:
Thanks! I make it print warning msg